lib/cdo/aws/cloud_formation.rb

```diff
@@ -0,0 +1,158 @@
+          puts "#{event.timestamp}- #{event.logical_resource_id} [#{event.resource_status}]: #{event.resource_status_reason}"
+        end
+        return
+      end
+      CDO.log.info "\nStack #{action} complete."
+    end
+
+    def self.delete
+      stack_exists = !!cfn.describe_stacks(stack_name: STACK_NAME) rescue false
+      if stack_exists
+        CDO.log.info "Shutting down #{STACK_NAME}..."
+        start_time = Time.now
+        cfn.delete_stack(stack_name: STACK_NAME)
+        wait_for_stack(:delete, start_time)
+      else
+        CDO.log.warn "Stack #{STACK_NAME} does not exist."
+      end
+    end
+
+    def self.create_or_update(cdn_enabled=false)
+      json_template = json_template(cdn_enabled)
+
+      update_certs
+      update_cookbooks
+      update_bootstrap_script
+
+      stack_exists = !!cfn.describe_stacks(stack_name: STACK_NAME) rescue false
+      action = stack_exists ? :update : :create
+      CDO.log.info "#{action} stack: #{STACK_NAME}..."
+      start_time = Time.now
+      updated_stack_id = cfn.method("#{action}_stack").call(
+        stack_name: STACK_NAME,
+        template_body: json_template,
+        capabilities: ['CAPABILITY_IAM']
+      ).stack_id
+      wait_for_stack(action, start_time)
+      CDO.log.info 'Outputs:'
+      cfn.describe_stacks(stack_name: updated_stack_id).stacks.first.outputs.each do |output|
+        CDO.log.info "#{output.output_key}: #{output.output_value}"
+      end
+    end
+
+    def self.validate(cdn_enabled=false)
+      json_template = json_template(cdn_enabled)
+      CDO.log.info JSON.pretty_generate(JSON.parse(json_template))
+      CDO.log.info cfn.validate_template(
+        template_body: json_template
+      ).description
+    end
+
+    def self.json_template(cdn_enabled)
+      template_string = File.read(aws_dir('cloudformation', 'cloud_formation_adhoc_standalone.yml.erb'))
+      @@local_variables = OpenStruct.new(
+        local_mode: !!CDO.chef_local_mode,
+        stack_name: STACK_NAME,
+        ssh_key_name: SSH_KEY_NAME,
+        image_id: IMAGE_ID,
+        instance_type: INSTANCE_TYPE,
+        branch: RakeUtils.git_branch,
+        region: CDO.aws_region,
+        environment: rack_env,
+        ssh_ip: SSH_IP,
+        certificate_arn: CERTIFICATE_ARN,
+        cdn_enabled: cdn_enabled,
+        domain: DOMAIN,
+        subdomain: FQDN,
+        availability_zone: Aws::EC2::Client.new.describe_availability_zones.availability_zones.first.zone_name,
+        file: method(:file)
+      )
+      YAML.load(ERB.new(template_string).result(@@local_variables.instance_eval{binding})).to_json
+    end
+
+    # Input filename, output ERB-processed file contents in CloudFormation JSON-compatible syntax (using Fn::Join operator).
+    def self.file(filename)
+      file = ERB.new(File.read(aws_dir('cloudformation', filename))).result(@@local_variables.instance_eval{binding})
```

Comment:
Consider adding an erb_eval helper function to factor out the duplicated code here and above for:

ERB.new(string).result(@@local_variables.instance_eval{binding})